### PR TITLE
Clean-up allowed types, and plugin config for TL3-XH & TL3-X

### PR DESCRIPTION
**Pull request by TonyMathiesen**
Changed allowed types matrix:
Current mapping assumptions for Growatt
SPF = SPF # not listed in MODBUS documentation under storage types, can maybe be combined? Storage (MIX, SPA, SPH) = GEN2
TL3-X (MAX, MID, MAC & MOD) = GEN
TL-X, TL-XH, TL-XHUS (MIN) = GEN3
TL3-XH (MOD) = GEN3

Clean up of GENx tags for the registers. GEN was not used, removed and used for TL3-X (non-hybrid) - except for MIN types, no change to that, as it uses same register sets as MOD TL3-XH. 

General:
Grid Export Limit, Limit Grid Export, Charger Switch moved up under Control.

For GEN3 (MOD TL3-XH and MIN types):
Changed many entities from input register 1-124 to 3000-3124. (same names and values) - maybe 1-2 changed name, needed to align names.  Added: Multiple EPS entities, Today's Yield, Total Yield

For GEN (TL3-X) non-hybrid:
Removed 3000-3124 register entities:
Total Forward Power
Total Reverse Power
Total Load Power
Today's Grid Import
Total Grid Import
Today's Grid Export
Total Grid Export
.. Have to check which is available for this model.

### CODE REVIEW

**Comment by TonyMathiesen** [comment]
@wills106 For issue #1057 
I have cleaned up the allowed types, and a lot of cleanup for GEN3 TL3-XH and MIN types. I forgot to mention that I removed GEN4, as those models should match with the GEN3. They should not be affected by the change. 

I have put it in this pull request, and then we can test 😉

File: custom_components/solax_modbus/plugin_growatt.py [hunk 1]
```diff
@@ -1677,16 +1677,16 @@ def value_function_combined_battery_power(initval, descr, datadict):
         register = 3049,
         scale = { 0: "Disabled",
                   1: "Enabled", },
-        allowedtypes = HYBRID | AC | GEN3 | GEN4,
```
Comment (wills106) [inline custom_components/solax_modbus/plugin_growatt.py:1680]: Need to check if AC should be left in for example the MIN TL-XA doesn't have PV. It's an AC only storage Inverter, so need a way to drop PV related entities on these Inverters.

File: custom_components/solax_modbus/plugin_growatt.py [hunk 2]
```diff
@@ -3439,22 +3406,22 @@ def value_function_combined_battery_power(initval, descr, datadict):
...
-#        unit = REGISTER_U8L,
+#        unit = REGISTER_U8L, #currently not working in the integration
 #        scale = { 0: "Standby", 1: "Normal", 3: "Fault", 4: "Flash", },
-#        allowedtypes = GEN3 | GEN4,
+#        allowedtypes = GEN3,
 #    ),
     GrowattModbusSensorEntityDescription(
-        name = "Total PV Power",
-        key = "total_pv_power",
+        name = "PV Power Total",
```
Comment (wills106) [inline custom_components/solax_modbus/plugin_growatt.py:3423]: Really need to be careful renaming entities, people will start crying.

File: custom_components/solax_modbus/plugin_growatt.py [hunk 3]
```diff
@@ -5023,15 +4996,18 @@ async def async_determineInverterType(self, hub, configdict):
             _LOGGER.error(f"{hub.name}: cannot find firmware version, even not for other Inverter")
             seriesnumber = "unknown"
 
+        ### Current mapping assumptions for Growatt
+        # SPF = SPF # not listed in MODBUS documentation under storage types, can maybe be combined?
```
Comment (wills106) [inline custom_components/solax_modbus/plugin_growatt.py:5000]: SPF uses a different set of Registers #686

**Review by wills106 — COMMENTED** [review]
Left a few General comments

File: custom_components/solax_modbus/plugin_growatt.py [hunk 4]
```diff
@@ -32,7 +32,7 @@
 ALL_X_GROUP    = X1 | X3
 
 PV             = 0x0400 # Needs further work on PV Only Inverters
-AC             = 0x0800
+AC             = 0x0800 # not used?
```
[the remark this answers is not in the document]
Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py:35]: Okay, we can leave it in. Could not see any "AC" on inverter type list. SPA was commented out.

File: custom_components/solax_modbus/plugin_growatt.py (hunk 3 above)
Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py:5000]: Noted. Comment removed. Nothing changed regarding SPF.

File: custom_components/solax_modbus/plugin_growatt.py [hunk 5]
```diff
@@ -50,7 +50,7 @@
 MPPT10         = 0x400000
 ALL_MPPT_GROUP = MPPT3 | MPPT4 | MPPT6 | MPPT8 | MPPT10
 
-ALLDEFAULT = 0 # should be equivalent to HYBRID | AC | GEN2 | GEN3 | GEN4 | X1 | X3
+ALLDEFAULT = 0 # should be equivalent to HYBRID | GEN | GEN2 | GEN3 | GEN4 | X1 | X3
```
[the remark this answers is not in the document]
Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py]: Adjusted to have both AC and GEN in the comment

File: custom_components/solax_modbus/plugin_growatt.py (hunk 2 above)
Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py:3423]: Agree. However having GEN4 just because of one entity name needs adjustment. 
I can maybe add a duplicate so it won't break any naming?
Only affected types was the single phase TL-X and TL-XH, where I have changed them to GEN3 as they uses same register sets.

File: custom_components/solax_modbus/plugin_growatt.py (hunk 1 above)
Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py:1680]: I don't see the AC being used by any inverter types.. Do I need to check anywhere else besides in the bottom of the file?

Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py:1680]: I can quickly add the AC in the ones where I removed it 😉

File: custom_components/solax_modbus/plugin_growatt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wills106) [reply custom_components/solax_modbus/plugin_growatt.py]: Commented out SPA is more a placeholder / reminder that they exist to my elf.

File: custom_components/solax_modbus/plugin_growatt.py (hunk 2 above)
Comment (wills106) [reply custom_components/solax_modbus/plugin_growatt.py:3423]: I didn't have a thorough look, if it's just the one perhaps if people rename the sensor to the new one before upgrading then they wont loose the history? Or have an alternative named one disabled by default using a value function to copy the values.

File: custom_components/solax_modbus/plugin_growatt.py (hunk 1 above)
Comment (wills106) [reply custom_components/solax_modbus/plugin_growatt.py:1680]: > I don't see the AC being used by any inverter types.. Do I need to check anywhere else besides in the bottom of the file?

There aren't any yet, more for future if someone with on starts using the Integration. Also AC only Inverters may still need further work to remove PV stuff?

File: custom_components/solax_modbus/plugin_growatt.py (hunk 2 above)
Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py]: Resolved. Reverted the name change, and simply added GEN3 to both. To avoid breaking for any of the models.

File: custom_components/solax_modbus/plugin_growatt.py (hunk 1 above)
Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py:1680]: Okay. Will re-add them a bit later today. So we are prepared 😉

Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py:1680]: Resolved

File: custom_components/solax_modbus/plugin_growatt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TonyMathiesen) [reply custom_components/solax_modbus/plugin_growatt.py]: Clear. Added them back 😉

**Comment by rmk-sys** [comment]
@wills106  & @TonyMathiesen,
thank you guys for your great work especially for the growatt plugin. It works great with my MOD 7000TL3-XH & ARK XH Battery System. If you like, I could do some testing on further improvements. A little improvement I can do today:

# register 3169 scale has to be set to 0.01

 register = 3169,
        register_type = REG_INPUT,
        scale = 0.01, #doc says 0.01, but datasheet of APX HV system module for MOD / MID TL-XH (BP) inverters have operating voltage range 600-980V.
        rounding = 2,
        allowedtypes = GEN3 | HYBRID,

**Comment by TonyMathiesen** [comment]
Hi @rmk-sys 😊
What battery voltage does the integration shows when the scale is 0.1? 
What voltage does your MOD inverter display in the Growatt Web UI or App? 
My APX HV shows, which also match with the datasheet. Which is why I changed it to scale 0.1: 
image
I have checked with the datasheet of ARK XH before changing it, and it operates at similar voltages as APX.

**Comment by rmk-sys** [comment]
@wills106  & @TonyMathiesen,
thank you guys for your great work especially for the growatt plugin. It works great with my MOD 7000TL3-XH & ARK XH Battery System. If you like, I could do some testing on further improvements. A little improvement I can do today:

# register 3169 scale has to be set to 0.01

 register = 3169,
        register_type = REG_INPUT,
        scale = 0.01, #doc says 0.01, but datasheet of APX HV system module for MOD / MID TL-XH (BP) inverters have operating voltage range 600-980V.
        rounding = 2,
        allowedtypes = GEN3 | HYBRID,

**Comment by TonyMathiesen** [comment]
@rmk-sys Strange. If I lookup the ARK XH battery system, which is the one that states "Compatible with MIN-XH, MOD-XH series inverter". This ARK XH operates at 400V for single phase and 650V for three phase independent of number of modules. Similar as APX HV. Needs scaling 0.01
The ARK HV operates around approx. 150-500 dependent of number of modules. Needs scaling 0.1

So it appears that for the same inverter models, there could be different scaling on the voltage depending on the battery system type. 
Maybe we should see if we can readout some information to detect if it one or the other battery system, and then handle the scaling accordingly. I can try to check if I can readout some meaningful information on my battery system.

**Comment by TonyMathiesen** [comment]
@rmk-sys What do you have under monitoring software version on your battery? 
image

**Comment by rmk-sys** [comment]
Hi again,

1.


software has been updated oct 1, but this is not the reason for the scaling issue. definitely the integration reads out reg. 3169 (GEN4 / Hybrid). probably your system reads out reg. 97, if you have no scaling issue. 

2.
I explored another scaling issue with grid power, which is nearly the same value as „output power": (for now I don’t know, which register my system is using and therefor I could not solve the problem.):



grid power L1-L3 have correct data and units. Not so „grid power“. (My system is definitely not using Grid Power VA and I don’t know why, while using L1-L3).

Is there reg. conflict? As you can see "Grid Power VA“ &  "Grid Power L1“ are using the same reg number.

    GrowattModbusSensorEntityDescription(
        name = "Grid Power VA",
        key = "grid_power_va",
        native_unit_of_measurement = UnitOfApparentPower.VOLT_AMPERE,
        device_class = SensorDeviceClass.APPARENT_POWER,
        state_class = SensorStateClass.MEASUREMENT,
        register = 3028,
        register_type = REG_INPUT,
        unit = REGISTER_U32,
        scale = 0.1,
        rounding = 1,
        allowedtypes = GEN4 | X1,
    ),
    GrowattModbusSensorEntityDescription(
        name = "Grid Power L1",
        key = "grid_power_l1",
        native_unit_of_measurement = UnitOfApparentPower.VOLT_AMPERE,
        device_class = SensorDeviceClass.APPARENT_POWER,
        state_class = SensorStateClass.MEASUREMENT,
        register = 3028,
        register_type = REG_INPUT,
        unit = REGISTER_U32,
        scale = 0.1,
        rounding = 1,
        allowedtypes = GEN4 | X3,

excuse me, but I’m not a programer, just the tester. ;-)

have a nice day, rainer

**Comment by TonyMathiesen** [comment]
@rmk-sys  No worries 😊

> 1. software has been updated oct 1, but this is not the reason for the scaling issue. definitely the integration reads out reg. 3169 (GEN4 / Hybrid). probably your system reads out reg. 97, if you have no scaling issue. 

I also use register 3169 for battery voltage. So there is a difference in scaling between the ARK HV system and the ARK XH + APX HV system. However I believe I have a solution for that. Could you share what your battery system is showing as "Monitoring Software Version", just to be sure it is different than "ZECA-3"? Because then I will use that to determining the battery system connected to set the scaling.

> I explored another scaling issue with grid power, which is nearly the same value as „output power": (for now I don’t know, which register my system is using and therefor I could not solve the problem.):
grid power L1-L3 have correct data and units. Not so „grid power“. (My system is definitely not using Grid Power VA and I don’t know why, while using L1-L3).

I'm not sure what you are asking. Could you rephrase it? 
You would have 3 "Grid Power L1,2,3" with unit VA and then you will have "Total Grid Power" and "Output" in W which should be the same, right? Please note that there will be a difference between "Total Grid Power" and "Total Load Power", as load will include the inverter and battery systems own consumption and grid will be negative when exporting. 

> Is there reg. conflict? As you can see "Grid Power VA“ &  "Grid Power L1“ are using the same reg number.

This is because single phase systems uses Grid Power VA and three phase systems uses Grid Power L1, L2, L3. The X1 or X3 in allowedtypes is selecting which to use.

**Comment by rmk-sys** [comment]
1. monitor version is ZEBA-5.

2. I'd like to have „grid_power_va“ with the right scaling (1617,5) and unit (VA), and not:



playing with that entity, didn’t succeed.

**Comment by TonyMathiesen** [comment]
@rmk-sys Please use the GitHub webpage if inserting pictures. Cannot see them if using mail. Please repost :)

**Comment by rmk-sys** [comment]
okay, sorry. I put it at #1057.

**Comment by TonyMathiesen** [comment]
Hi @rmk-sys 
Now I'm a bit confused. What exact model inverter do you have? Please verify that it is a MOD 7000 TL3-XH, meaning 3-phase. 
Your picture displays "grid_power_va" which is for single phase inverter. What is the integration showing under "Firmware Version"?

**Comment by rmk-sys** [comment]
yes, it’ a MOD7000 TL3-XH 3 phase. latest integration 2024.10.3. Seriennummer: DN1.0ZBDB (from integration). 
DN1.0/DNAA065100/ZBDB-0008/S07B00D00T00P0FU01M0046 (from web ui).

(by the way PR #1006 seems to be closed)

hope, i get, what you need. sorry for inconvenience, rainer

**Comment by rmk-sys** [comment]
entity serialnumber returns nothing. 
i found just that:
Bildschirmfoto 1

**Comment by TonyMathiesen** [comment]
@rmk-sys Thanks for confirming. Then we have the same type inverter. Mine is a MOD 8K TL3-XH. Also DN1.0
I believe I found the issue. Was running a local version as I'm trying to add some stuff on our inverter. Copied the 2024.10.3, and can see that "Total Grid Power" is running with a missing scale 0.1 

I have created a PR to fix it :) #1098 
image

If you want to check now, you can change this in plugin_growatt.py:
image

**Comment by TonyMathiesen** [comment]
Forgot to mention that this is with the unit of Watts, as this is according to the documentation from Growatt. 

Do you want to have a "Total Grid Power VA" which is equal to Grid Power L1 (VA) + Grid Power L2 (VA) + Grid Power L3 (VA)?

**Comment by TonyMathiesen** [comment]
Like so: 
image

I believe it makes sense to change, as "Total Grid Power" (Watts) is actually reading 3023 which is the Output Power (W), and it is the exact same as "Output Power".

**Comment by rmk-sys** [comment]
thx for now, i changed it as you mentioned:
Bildschirmfoto

but the addition of the 3 phases wil not resume the total grid power in va. i'am thinking by myself, that there must be a register to get total output in va, cause the inverters display shows it.
but all this are minor tweaks, which don't affect me very hard. i just tested it for the improvement of the integration.

what i'd really like to have is controlling the priority and inverters state (standby/on grid) which is recommended for changing PF factor/PF curve and other changes.

again, thanks for all the work, you and willis invested in this 99% great working integration, rainer

**Comment by TonyMathiesen** [comment]
The (Netzleistung) "Total Grid Power" will be changed to VA with PR #1098 .. Because it makes really good sense to have it like this. As "Output Power" (W) is already there. 

The battery scaling adjust based on which battery system is connected will be fixed with PR #1099. That will correct your battery voltage. 

> what i'd really like to have is controlling the priority and inverters state (standby/on grid) which is recommended for changing PF factor/PF curve and other changes.

The priority is what I've been working on adding here on my system for the last two weeks or so - so it is coming ;) 
Regarding inverter state, there I'm not sure. Is that something you are changing in the App or Web UI at the moment?

So it is coming when the PRs are merged into a new version :) 

> again, thanks for all the work, you and willis invested in this 99% great working integration, rainer

It's really kudos to @wills106 as he is the one behind this integration. I have only done some tweaking for the Growatt inverters over the past two months, since I got my system. Really a cool integration :)

**Comment by TonyMathiesen** [comment]
Hi rainer @rmk-sys 
Check out version 2024.10.4. The battery voltage scaling, grid power, etc is included in that version. Together with time slot 1, 2, 3 for priority settings.

**Comment by rmk-sys** [comment]
@TonyMathiesen,
hi again. for the issues, we discussed: everything looks fine! battery, grid toal va firmwere control number and serialnumber, perfect. the controls I'm going to test within the next days. thx

**Comment by rmk-sys** [comment]
@TonyMathiesen, @wills106

Hello Tony,
I’ve been very busy the last 2 weeks. Since Nov.1 I’m using my MIC 3600 TL-XH, cause this device is more energy efficient during the low light months. Most of the data are displayed fine even. Although there are some little glitches I explored:

1. FW version: afaik the complete FW has to be for example: AL1.0/ALBA090101
   wouldn’t it be nice to have here the whole version number?
the FW Control Version (afaik it’s the FW communication Version) is displayed well: ZABA-0018.

2. I’m a bit confused about Inverter State (sensor.inverter_state). both of my inverters are displaying all the time ‚waiting‘ although web ui (System Status) says ‚on grid mode‘. and what  (register) is ‚sensor.run_mode’ which displays ‚normal‘.

I found this:
Bildschirmfoto 1
if it’s ‚on grid/standby‘ it would be nice to have a control entity.

3. Device information displays FW version and not serial number.
Bildschirmfoto
best regards, rainer

P.S: Can I buy You a coffee?
